### ISSUE TITLE
Modifying argmax search to beam search

### ISSUE THREAD

**Issue by ghost** [mmehdig/lm_beam_search#1]
I have a keras version of code where the prection is as follows
           
` predictions = np.argmax(model.predict(X), axis=2)` where `X` is taken from the input from the source file as follows

```
def load_test_data(source, X_word_to_ix, max_len):
    f = open(source, 'r')
    X_data = f.read()
    f.close()
    X = [text_to_word_sequence(x)[::-1] for x in X_data.split('\n') if len(x) > 0 and len(x) <= max_len]
    for i, sentence in enumerate(X):
        for j, word in enumerate(sentence):
            if word in X_word_to_ix:
                X[i][j] = X_word_to_ix[word]
            else:
                X[i][j] = X_word_to_ix['UNK']
    return X
```

Can you please help me what modifying your function to my input format ? It is difficult to understand what each of src_input and model mean.Please add a README with detailed description

**Comment by mmehdig** [comment]
Sure, I'll try my best. 

My model is a somthing2seq translation model. So, the input `X` is in fact: `X = [srcs, trgs]`. Each item in `srcs` has a parallel in `trgs`. We initiate the sentence with start tag and dummy zeros but on each step prediction the `trgs` changes along the way.  The model predicts all the way together but I only took the slice which we have in search. For example in translation model from French to English, if the source is a sequence of three words: `['un', 'deux', 'trois']`, when predicting the target sequence at first we only have an empty sequence: `['<S>', '<PAD>', '<PAD>', '<PAD>']` then on each step of prediction we update the target:
```
['<S>', 'one', '<PAD>', '<PAD>']
['<S>', 'one', 'two', '<PAD>']
['<S>', 'one', 'two', 'three']
```
Note: if your model doesn't need to update the targets, then maybe you don't need beam search.

In the current version of the code, `src_input` can be in any shape depending on your model, and the target sequence we are searching for is `sent_predict`. Both `src_input` and `sent_predict` are inputs of the model. 

On the target word prediction, I encoded both `'<S>'` and `'<PAD>'` with zero. If you want to adapt my code you probably need to change these parts:

In line 5 (initialization):
```
k_beam = [(0, [0]*(sequence_max_len+1))]
```
to
``` 
k_beam = [(0, [X_word_to_ix['<S>']]+[X_word_to_ix['<PAD>']]*(sequence_max_len+1))]
```
also, line 19 (update):
```
list(sent_predict[:l+1])+[next_wid]+[0]*(sequence_max_len-l-1)
```
to 
```
list(sent_predict[:l+1])+[next_wid]+[X_word_to_ix['<PAD>']]*(sequence_max_len-l-1)
```

Also in the case, that encoding of words in X_word_to_ix doesn't start from zero and the zero palce in sortmax output needs an adjustment to be mapped on a meaningful word encoding you need to change line 13 as well: 
```
possible_k = predicted[l].argsort()[-k:][::-1]
```
to 
```
possible_k = predicted[l].argsort()[-k:][::-1] + adjustment?
```

**Comment by ghost** [comment]
Nice explanation @mmehdig  and i am using this implementation

https://github.com/ChunML/seq2seq 

In this implementation it has argmax search and i was trying to add your beam search.It would be great if you can help me integrate this.I am unable to do so

Thanks in advance

**Comment by mmehdig** [comment]
Well, I cannot fix someone else's model. But I can point out the problem with that implementations. 

Here is the code I found in that repository:
```
def create_model(X_vocab_len, X_max_len, y_vocab_len, y_max_len, hidden_size, num_layers):
    model = Sequential()

    # Creating encoder network
    model.add(Embedding(X_vocab_len, 1000, input_length=X_max_len, mask_zero=True))
    model.add(LSTM(hidden_size))
    model.add(RepeatVector(y_max_len))

    # Creating decoder network
    for _ in range(num_layers):
        model.add(LSTM(hidden_size, return_sequences=True))
    model.add(TimeDistributed(Dense(y_vocab_len)))
    model.add(Activation('softmax'))
    model.compile(loss='categorical_crossentropy',
            optimizer='rmsprop',
            metrics=['accuracy'])
    return model
```

This model doesn't follow the standard seq2seq model. The decoder in this case is not using the predicted output as input to the next prediction. Only the hidden state if the LSTM is past to the next step.

As he mentioned in his blog, the standard seq2seq model roughly is in this shape:

![](https://chunml.github.io/images/projects/sequence-to-sequence/figure.png)

The outputs are built in a loop:
```
['A', 'B', 'C', '<eos>', '<eos>', '<eos>', '<eos>', '<eos>'] =>  ['W', '<eos>', '<eos>', '<eos>', '<eos>'] 
['A', 'B', 'C', '<eos>', 'W', '<eos>', '<eos>', '<eos>']  => ['W', 'X', '<eos>', '<eos>', '<eos>'] 
['A', 'B', 'C', '<eos>', 'W', 'X', '<eos>', '<eos>']  => ['W', 'X', 'Y', '<eos>', '<eos>'] 
['A', 'B', 'C', '<eos>', 'W', 'X', 'Y', '<eos>']  = ['W', 'X', 'Y', 'Z', '<eos>'] 
['A', 'B', 'C', '<eos>', 'W', 'X', 'Y', 'Z'] => ['W', 'X', 'Y', 'Z', '<eos>'] 
```

But in his model, the outputs are built at once as you pass the source input:

```
['A', 'B', 'C']  => ['W', 'X', 'Y', 'Z', '<eos>'] 
```
From the [blog post](https://chunml.github.io/ChunML.github.io/project/Sequence-To-Sequence/):
![](https://chunml.github.io/images/projects/sequence-to-sequence/repeated_vector.png)

This doesn't need beam search, and it is fine if it works :)

**Comment by ghost** [comment]
Now I get it ! Thanks for the clarification.


I have gone through your model as well and as mentioned the attention is messed up.In case you are making further changes in your repository,please test it on europarl or some publicly available large parallel corpus and include the attention separately with and without in the decoder and obviously a well documented readme
